### PR TITLE
Fastify support

### PR DESCRIPTION
**Pull request by juliangoacher**
* *Fastify support*: This PR adds a new plugin package, `trail-fastify-plugin` based on Fastify v2.12.0. The plugin uses `fluent-schema` for JSON schema support. The plugin reproduces nearly exactly the behaviour of `trail-hapi-plugin` except when validating HTTP headers. The `trail-fastify-server` package provides a minimal Fastify server pre-configured to use `trail-fastify-plugin`.
* *Fix swagger support in Hapi plugin*: The Swagger documentation for `trail-hapi-plugin` was broken after the update to the latest Hapi and Joi versions. This was due to incorrect JSON schema generation in the Swagger spec doc, and has been fixed by using `joi-to-swagger`.
* *Update minimum node version in sub-packages*: The required node version specified in the `package.json` of the core and Hapi sub-packages still referenced node v8; this has been updated to node v12.
* *Upgrade inert dependency in trail-hapi-server*: As part of npm audit fix.
* *Upgrade boom dependency to latest available*
* *No change to module versions*: The module and all sub-package versions are kept at `3.0.0`.

### CODE REVIEW

File: packages/trail-fastify-plugin/README.md [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+```
+npm install @nearform/trail-fastify-plugin
+```
+
+## Usage
+
+```javascript
+const main = async function() {
+  import Fastify from 'fastify'
```
Comment (mcollina) [inline packages/trail-fastify-plugin/README.md]: This is not valid JavaScript. Please use require at the top of the file.

File: packages/trail-fastify-plugin/README.md [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  const fastify = Fastify()
+
+  fastify.register(require('@nearform/trail-fastify-plugin'))
+
+  await fastify.ready()
```
Comment (mcollina) [inline packages/trail-fastify-plugin/README.md]: this `await fastify.ready()` call is not needed

File: packages/trail-fastify-plugin/lib/index.js [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+async function trail (server, options) {
+  const trailsManager = new TrailsManager(undefined, options.pool)
+
+  server.decorate('trailCore', trailsManager)
+  server.decorateReply('trailCore', trailsManager)
+
+  server.addHook('onClose', async (instance, done) => {
+    await trailsManager.close()
+    done()
```
Comment (mcollina) [inline packages/trail-fastify-plugin/lib/index.js]: remove `done` here and in the function signature.

File: packages/trail-fastify-plugin/lib/index.js [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+  })
+
+  const ajv = new Ajv({ allErrors: true })
```
Comment (mcollina) [inline packages/trail-fastify-plugin/lib/index.js]: why is this needed?

File: packages/trail-fastify-plugin/lib/routes/trails.js [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+    })
+  }
+
+  addApiRoute(fastify, 'trails', {
+    method: 'GET',
+    path: '/trails',
+    async handler (request, reply) {
+      const { from, to, who, what, subject, page, pageSize, sort } = request.query
+      const results = await reply.trailCore.search({ from, to, who, what, subject, page, pageSize, sort })
+      reply.send(results)
```
Comment (mcollina) [inline packages/trail-fastify-plugin/lib/routes/trails.js]:
```suggestion
      return results
```

File: packages/trail-fastify-plugin/lib/routes/trails.js [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+    }
+  })
+
+  addApiRoute(fastify, 'trails', {
+    method: 'GET',
+    path: '/trails/enumerate',
+    async handler (request, reply) {
+      const { from, to, type, page, pageSize, desc } = request.query
+      const results = await reply.trailCore.enumerate({ from, to, type, page, pageSize, desc })
+      reply.send(results)
```
Comment (mcollina) [inline packages/trail-fastify-plugin/lib/routes/trails.js]:
```suggestion
      return results
```

File: packages/trail-fastify-plugin/lib/routes/trails.js [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+    }
+  })
+
+  addApiRoute(fastify, 'trails', {
+    method: 'POST',
+    path: '/trails',
+    async handler (request, reply) {
+      const id = await reply.trailCore.insert(request.body)
+      const trail = await reply.trailCore.get(id)
+      reply.code(201).send(trail)
```
Comment (mcollina) [inline packages/trail-fastify-plugin/lib/routes/trails.js]:
```suggestion
      reply.code(201)
      return trail
```

File: packages/trail-fastify-plugin/lib/routes/trails.js [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+  addApiRoute(fastify, 'trails', {
+    method: 'GET',
+    path: '/trails/:id',
+    async handler (request, reply) {
+      const { id } = request.params
+      const trail = await reply.trailCore.get(id)
+
+      if (!trail) throw notFound(`Trail with id ${id} not found.`)
+
+      return reply.send(trail)
```
Comment (mcollina) [inline packages/trail-fastify-plugin/lib/routes/trails.js]:
```suggestion
      return trail
```

File: packages/trail-fastify-plugin/lib/routes/trails.js [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+    method: 'PUT',
+    path: '/trails/:id',
+    async handler (request, reply) {
+      const { id } = request.params
+      const updated = await reply.trailCore.update(id, request.body)
+
+      if (!updated) throw notFound(`Trail with id ${id} not found.`)
+
+      const trail = await reply.trailCore.get(id)
+      reply.code(202).send(trail)
```
Comment (mcollina) [inline packages/trail-fastify-plugin/lib/routes/trails.js]:
```suggestion
      reply.code(202)
      return send(trail)
```

File: packages/trail-fastify-plugin/lib/routes/trails.js [hunk 10]
```diff
@@ -0,0 +1,184 @@
+const { notFound } = require('@hapi/boom')
+const S = require('fluent-schema')
+
+const { errorsSchemas } = require('../schemas/errors')
+const { spec, trailSchema } = require('../schemas/trails')
+const { addApiRoute, generateSpec } = require('../api')
+
+module.exports = async function (fastify, options, done) {
```
Comment (mcollina) [inline packages/trail-fastify-plugin/lib/routes/trails.js]:
```suggestion
module.exports = async function (fastify, options) {
```

File: packages/trail-fastify-plugin/lib/routes/trails.js [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+      auth: false,
+      description: 'Delete a audit trail.',
+      tags: ['api', 'trails']
+    }
+  })
+
+  // Add tagged routes to the swagger.json
+  generateSpec(spec, 'trails')
+
+  done()
```
Comment (mcollina) [inline packages/trail-fastify-plugin/lib/routes/trails.js]:
```suggestion
```

File: packages/trail-fastify-plugin/lib/schemas/errors.js [hunk 12]
```diff
@@ -0,0 +1,64 @@
+const S = require('fluent-schema')
```
Comment (mcollina) [inline packages/trail-fastify-plugin/lib/schemas/errors.js]: add use strict at the top.

File: packages/trail-fastify-plugin/lib/index.js (hunk 4 above)
Comment (juliangoacher) [reply packages/trail-fastify-plugin/lib/index.js]: So that all of Ajv's configuration options are easily available. Wasn't clear at start which ones I would need to change, and left this in as may be useful later in the project to change the settings.

**Comment by juliangoacher** [comment]
The Fastify plugin uses the exact same unit tests as the Hapi plugin, except for one unit test for invalid HTTP headers.

**Comment by mcollina** [comment]
> The Fastify plugin uses the exact same unit tests as the Hapi plugin, except for one unit test for invalid HTTP headers.

How are those hooked up? Can you send a couple of links?

**Comment by juliangoacher** [comment]
@mcollina I mean just that the unit tests under https://github.com/nearform/trail/tree/15-fastify-support/packages/trail-fastify-plugin/test are copied from trails-hapi-plugin with one exception (the unit test for an invalid content-type header). Same for the fastify server code.

File: packages/trail-fastify-plugin/lib/api.js [hunk 13]
```diff
@@ -0,0 +1,115 @@
+const { get } = require('lodash')
```
Comment (mcollina) [inline packages/trail-fastify-plugin/lib/api.js]: use strict is missing.

What's this file purpose, I'm a bit lost.

**Comment by mcollina** [comment]
CI is failing, would you mind you fixing it?

**Comment by juliangoacher** [comment]
@mcollina CI is failing because travis can't find the trail-fastify-plugin & trail-fastify-server modules on npm - I'm not really sure how to fix this without publishing the two modules.

File: packages/trail-fastify-plugin/lib/api.js (hunk 13 above)
Comment (juliangoacher) [reply packages/trail-fastify-plugin/lib/api.js]: The general project structure, including this file, is copied from the trail-hapi-plugin code. The function of the file is to support generation of the swagger spec from the api route definitions. Note that its functions are also called by the corresponding server module.

**Comment by juliangoacher** [comment]
@mcollina CI build is passing now; turns out lerna had added the sub-packages as direct dependencies of the root module.

File: packages/trail-fastify-plugin/lib/index.js [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+        stack
+      }
+    }
+    reply.code(code).send(error)
+  })
+
+  await server.register(require('./routes/trails'))
+}
+
+module.exports = fp(trail, { name: 'trail-fastify-plugin' })
```
[the remark this answers is not in the document]
Comment (juliangoacher) [reply packages/trail-fastify-plugin/lib/index.js]: I didn't know about fastify-swagger, I'll take a look at it now.
